modules/3_detection/src/detector.cpp

```diff
@@ -1,28 +1,161 @@
 #include "detector.hpp"
-
+#include <utility> 
 #include <opencv2/opencv.hpp>
 #include <opencv2/core/utils/filesystem.hpp>
 #include <inference_engine.hpp>
+#include <algorithm>
 
 using namespace cv;
 using namespace InferenceEngine;
+using namespace cv::utils::fs;
 
 Detector::Detector() {
     Core ie;
 
     // Load deep learning network into memory
     auto net = ie.ReadNetwork(utils::fs::join(DATA_FOLDER, "face-detection-0104.xml"),
                               utils::fs::join(DATA_FOLDER, "face-detection-0104.bin"));
+	// Specify preprocessing procedures
+	// (NOTE: this part is different for different models!)
+	InputInfo::Ptr inputInfo = net.getInputsInfo()["image"];
+	inputInfo->getPreProcess().setResizeAlgorithm(ResizeAlgorithm::RESIZE_BILINEAR);
+	inputInfo->setLayout(Layout::NHWC);
+	inputInfo->setPrecision(Precision::U8);
+	outputName = net.getOutputsInfo().begin()->first;
+
+	// Initialize runnable object on CPU device
+	ExecutableNetwork execNet = ie.LoadNetwork(net, "CPU");
+
+	// Create a single processing thread
+	req = execNet.CreateInferRequest();
+}
+
+Blob::Ptr wrapMatToBlob(const Mat& m) {
+	CV_Assert(m.depth() == CV_8U);
+	std::vector<size_t> dims = { 1, (size_t)m.channels(), (size_t)m.rows, (size_t)m.cols };
+	return make_shared_blob<uint8_t>(TensorDesc(Precision::U8, dims, Layout::NHWC),
+		m.data);
 }
 
 void Detector::detect(const cv::Mat& image, std::vector<cv::Rect>& boxes,
                       std::vector<float>& probabilities,
-                      std::vector<unsigned>& classes) {
-    CV_Error(Error::StsNotImplemented, "detect");
+                      std::vector<unsigned>& classes, const float nmsThreshold, 
+					  const float probThreshold) {
+	// Create 4D blob from BGR image
+	Blob::Ptr input = wrapMatToBlob(image);
+
+	req.SetBlob("image", input);
+
+	// Launch network
+	req.Infer();
+	auto w = image.cols;
+	auto h = image.rows;
+
+	// Copy output. 
+	float* output = req.GetBlob(outputName)->buffer();
+	int size = req.GetBlob(outputName)->size();
+	float score = 0;
+	float cls = 0;
+	float id = 0;
+	for (int i = 0; i < size/7; i++)
+	{
+		score = output[i * 7 + 2];
+		cls = output[i * 7 + 1];
+		id = output[i * 7];
+		if (id >= 0 && score > probThreshold && cls <= 1)
```

Comment:
Are you sure that we need classes with `cls=0`?

Reply:
No, but where can I find information about the meaning of classes ids?

Reply:
I think we will add one more example here. Actually, it's just an index for multiclass detection. In case of face detection there is only one class - face.